wit/window.wit

```diff
@@ -0,0 +1,228 @@
+       brightness-down,
+       brightness-up,
+       print,
+       
+       f1,
+       f2,
+       f3,
+       f4,
+       f5,
+       f6,
+       f7,
+       f8,
+       f9,
+       f10,
+       f11,
+       f12,
+       f13,
+       f14,
+       f15,
+       f16,
+       f17,
+       f18,
+       f19,
+       f20,
+       
+       app-switch,
+       
+       channel-down,
+       channel-up,
+       fast-forward,
+       media-pause,
+       media-play,
+       media-play-pause,
+       media-record,
+       media-rewind,
+       media-stop,
+       media-track-next,
+       media-track-previous,
+       
+       audio-volume-down,
+       audio-volume-up,
+       audio-volume-mute,
+    }
+    
+    flags modifiers {
+        ctrl,
+        alt,
+        alt-graph,
+        shift,
+        caps-lock,
+        meta,
+    }
+    
+    /// Describes the key location on the keyboard, for keys that appear more than once on a keyboard.
+    enum key-location {
+        standard,
+        left,
+        right,
+        numpad,
+    }
+    
+    variant key {
+        /// The key has an associated Unicode scalar value.
+        character(u32),
+        /// The key is a key without a Unicode scalar value.
+        special(special-key),
+        /// A dead key was pressed. The next key event will contain the combined character.
+        dead,
+        /// The key could not be identified.
+        unidentified,
+    }
+    
+    
+    record key-data {
+        modifiers: modifiers,
+        key: key,
+        code: key-code,
+        location: key-location,
+        repeat: bool,
+    }
+    
+    flags mouse-buttons {
+        left,
+        middle,
+        right,
+    }
+    
+    record mouse {
+        position: position,
+        buttons: mouse-buttons,
+    }
+    
+    record scroll {
+        mouse: mouse,
+        x: f32,
+        y: f32,
+    }
+    
+    
+    /// Window system capabilities that are supported.
+    /// Methods that use unsupported capabilities will trap in WASM.
+    flags capabilities {
+        /// The windows have a title that can be set.
+        title,
+        /// The windows have position on a display which can be moved by the program. Not given e.g. on the web, where position is handled by CSS.
+        position,
+        // The windows have icons which can be set by the program.
+        icon,
+        /// Windows can request to go into fullscreen mode. This capability not being present can mean fullscreen windows
+        /// are not supported, or the window is forced fullscreen.
+        fullscreen,
+    }
+    
+    
+    record position {
+        x: u16,
+        y: u16,
+    }
+    
+    record size {
+        width: u16,
+        height: u16,
+    }
+    
+    record create-desc {
+        title: string,
+        position: position,
+        size: size,
+        visible: bool,
+    }
+    
+    /// Checks whether window functionality is available at all in the runtime.
+    available: func() -> bool;
```

Comment:
What does this mean? Why would you even be able to get here, when there's no windowing capabilities?

Reply:
Yeah. You talked about the possibility of having multiple presentation APIs, and I also added an API intended for raw framebuffer access on connected screens, for embedded devices. If this were Rust, I'd just make the window constructor return an option, but I don't think that's possible in WIT?

Reply:
Wit does support options: e.g. `option<bool>` instead of `bool`.

But I'm not sure I'm following why this is necessary. If another presentation api is used, how would you reach this point at all?
You would use `wasi-windowing-other` instead of `wasi-windowing`, and `wasi-windowing-other` would have a completely different set of functions.

Reply:
Currently the unit of implementation is a world. So if something is included in the world, you have to implement it, even if it's stubbed. Since I can't make the constructor fail gracefully, there needs to be a function to query runtime support. Of course each display API could get its own world, too.

Reply:
> If this were Rust, I'd just make the window constructor return an option, but I don't think that's possible in WIT?

The way to do that in WIT is to use a static factory function:
```
ctor: static func([...]) -> option<the-type>;
```

Reply:
> Of course each display API could get its own world, too.

That's what I'd prefer.